src/crimson/os/seastore/segment_cleaner.cc

```diff
@@ -0,0 +1,100 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "crimson/common/log.h"
+
+#include "crimson/os/seastore/segment_cleaner.h"
+
+namespace {
+  seastar::logger& logger() {
+    return crimson::get_logger(ceph_subsys_filestore);
+  }
+}
+
+namespace crimson::os::seastore {
+
+SegmentCleaner::get_segment_ret SegmentCleaner::get_segment()
+{
+  // TODO
+  return get_segment_ret(
+    get_segment_ertr::ready_future_marker{},
+    next++);
+}
+
+void SegmentCleaner::update_journal_tail_target(journal_seq_t target)
+{
+  logger().debug(
+    "{}: {}",
+    __func__,
+    target);
+  assert(journal_tail_target == journal_seq_t() || target >= journal_tail_target);
+  if (journal_tail_target == journal_seq_t() || target > journal_tail_target) {
+    journal_tail_target = target;
+  }
+}
+
+void SegmentCleaner::update_journal_tail_committed(journal_seq_t committed)
+{
+  if (journal_tail_committed == journal_seq_t() ||
+      committed > journal_tail_committed) {
+    logger().debug(
+      "{}: update journal_tail_committed {}",
+      __func__,
+      committed);
+    journal_tail_committed = committed;
+    }
+  if (journal_tail_target == journal_seq_t() ||
+      committed > journal_tail_target) {
```

Comment:
```suggestion
  if (committed > journal_tail_target) {
```
i think this would suffice. because, `journal_tail_target()` returns an aggregated initialized `journal_seq_t` whose member variables are all in turn aggregated initialized, and since all members are `unsigned`, so they are zero.

so, `committed > journal_tail_target` also covers the case where  `journal_tail_target` is "empty", unless `committed` could be "empty" as well.

Reply:
paddr_t has a constructor with the segment and offset set to NULL_SEG_ID and NULL_SEG_OFF, so the sorting doesn't happen to work out.